File: silencendo/src/llm/deepseek.ts
import { LLMClient, LLMMessage } from './types';
import { Chunk } from '../ingestion/types';
import { Context } from '../context/types';
import OpenAI from 'openai';
import 'dotenv/config';

// Type alias to match OpenAI's expected format
type OpenAIMessage = {
  role: 'user' | 'assistant' | 'system';
  content: string;
};

export class DeepSeekLLM implements LLMClient {
  private client: OpenAI;
  private model: string;

  constructor() {
    const apiKey = process.env.DEEPSEEK_API_KEY;
    if (!apiKey) {
      throw new Error(
        'DEEPSEEK_API_KEY environment variable is required. ' +
        'Please set it using: export DEEPSEEK_API_KEY=your_key_here'
      );
    }

    const baseUrl = process.env.DEEPSEEK_BASE_URL || 'https://api.deepseek.com';
    this.model = process.env.DEEPSEEK_MODEL || 'deepseek-chat';

    this.client = new OpenAI({
      apiKey,
      baseURL: baseUrl,
    });
  }

  async generate(messages: LLMMessage[]): Promise<string> {
    try {
      // Convert our LLMMessage type to OpenAI's expected format
      const openAiMessages: OpenAIMessage[] = messages.map(msg => ({
        role: msg.role,
        content: msg.content
      }));

      const response = await this.client.chat.completions.create({
        model: this.model,
        messages: openAiMessages,
        temperature: 0.7,
      });

      return response.choices[0]?.message?.content || '';
    } catch (error) {
      throw new Error(`DeepSeek API error: ${(error as Error).message}`);
    }
  }

  async answer(question: string, retrievedChunks: Chunk[], context: Context): Promise<string> {
    // If no chunks are found, let DeepSeek answer the question based on its general knowledge
    // but provide context about the project
    if (retrievedChunks.length === 0) {
      const systemMessage = `You are an AI assistant. The user asked a question but no specific context was provided from their documents.

PROJECT CONTEXT:
Project: ${context.project || 'Not set'}
Stage: ${context.stage || 'Not set'}
Task: ${context.task || 'Not set'}

If the question is related to the project context above, provide a relevant response.
If the question is general and not related to the project, you may use your general knowledge to answer.
If the question is specific to the user's documents and you have no context, state that you don't have the relevant information from the documents.`;

      const userMessage = `Question: ${question}

Please provide an appropriate response based on the project context and your general knowledge if applicable.`;

      try {
        const response = await this.client.chat.completions.create({
          model: this.model,
          messages: [
            { role: 'system', content: systemMessage },
            { role: 'user', content: userMessage }
          ],
          temperature: 0.7,
        });

        return response.choices[0]?.message?.content || 'No response generated.';
      } catch (error) {
        return `DeepSeek API error: ${(error as Error).message}`;
      }
    }

    // Format the chunks for the system prompt when chunks are available
    const formattedChunks = retrievedChunks
      .map((chunk, index) => `[SOURCE ${index + 1}]: ${chunk.text}`)
      .join('\n\n');

    const systemMessage = `You are an AI assistant that answers questions based ONLY on the provided CONTEXT CHUNKS.
If the answer is not found in the provided chunks, explicitly state that you cannot answer from the provided sources.
Do not fabricate information or rely on general knowledge when in strict mode.

CONTEXT CHUNKS:
${formattedChunks}

PROJECT CONTEXT:
Project: ${context.project || 'Not set'}
Stage: ${context.stage || 'Not set'}
Task: ${context.task || 'Not set'}

Follow these rules:
1. Answer ONLY using information from the CONTEXT CHUNKS
2. If information is not in the chunks, say you cannot answer from sources
3. Reference the source number when citing information
4. Be concise and accurate`;

    const userMessage = `Question: ${question}

Please provide a comprehensive answer based on the provided context chunks. If the information is not available in the provided sources, state that you cannot answer from the provided sources.`;

    try {
      const response = await this.client.chat.completions.create({
        model: this.model,
        messages: [
          { role: 'system', content: systemMessage },
          { role: 'user', content: userMessage }
        ],
        temperature: 0.3,
      });

      return response.choices[0]?.message?.content || 'No response generated.';
    } catch (error) {
      return `DeepSeek API error: ${(error as Error).message}`;
    }
  }

  async edit(inputText: string, instruction: string, retrievedChunks: Chunk[], context: Context): Promise<string> {
    // Format the chunks for context
    const formattedChunks = retrievedChunks
      .map((chunk, index) => `[SOURCE ${index + 1}]: ${chunk.text}`)
      .join('\n\n');

    const systemMessage = `You are an AI editor that modifies text based on specific instructions.
Use the provided CONTEXT CHUNKS as reference material when applicable to the editing task.
Apply the given instruction to transform the input text while maintaining accuracy and coherence.

CONTEXT CHUNKS:
${formattedChunks}

PROJECT CONTEXT:
Project: ${context.project || 'Not set'}
Stage: ${context.stage || 'Not set'}  
Task: ${context.task || 'Not set'}

Editing Rules:
1. Follow the instruction precisely
2. Use context chunks as reference when relevant to the edit
3. Preserve the original meaning unless instructed otherwise
4. Make minimal changes unless specifically asked for significant modifications`;

    const userMessage = `Original Text: ${inputText}

Instruction: ${instruction}

Please apply the instruction to the original text. Return only the edited text without additional commentary.`;

    try {
      const response = await this.client.chat.completions.create({
        model: this.model,
        messages: [
          { role: 'system', content: systemMessage },
          { role: 'user', content: userMessage }
        ],
        temperature: 0.5,
      });

      return response.choices[0]?.message?.content || 'No edit generated.';
    } catch (error) {
      return `DeepSeek API error: ${(error as Error).message}`;
    }
  }
}

File: silencendo/llm/deepseek.go
package llm

import (
	stdcontext "context"
	"fmt"
	"os"
	"strings"

	"silencendo/context"
	"silencendo/ingestion"

	"github.com/sashabaranov/go-openai"
)

type DeepSeekLLM struct {
	client *openai.Client
	model  string
}

const groundedAnswerSystem = "You are a retrieval QA agent. Use ONLY the provided context. If the context is empty or missing the answer, reply exactly: 'I don't have enough information to answer that yet. Please add sources or ingest them.' Do not invent names, people, or facts."

const groundedEditSystem = "You are an editing assistant. Use ONLY the provided context plus the input text. Do not invent names, people, or facts. If context is empty, operate only on the input text and instruction without adding new factual details."

func NewDeepSeekLLM() (*DeepSeekLLM, error) {
	apiKey := strings.TrimSpace(os.Getenv("DEEPSEEK_API_KEY"))
	if apiKey == "" {
		return nil, fmt.Errorf("DEEPSEEK_API_KEY is required")
	}

	baseURL := strings.TrimSpace(os.Getenv("DEEPSEEK_BASE_URL"))
	if baseURL == "" {
		baseURL = "https://api.deepseek.com"
	}

	model := strings.TrimSpace(os.Getenv("DEEPSEEK_MODEL"))
	if model == "" {
		model = "deepseek-chat"
	}

	config := openai.DefaultConfig(apiKey)
	config.BaseURL = baseURL

	client := openai.NewClientWithConfig(config)

	return &DeepSeekLLM{
		client: client,
		model:  model,
	}, nil
}

func (d *DeepSeekLLM) Generate(ctx stdcontext.Context, messages []Message) (string, error) {
	openaiMessages := make([]openai.ChatCompletionMessage, len(messages))
	for i, msg := range messages {
		openaiMessages[i] = openai.ChatCompletionMessage{
			Role:    msg.Role,
			Content: msg.Content,
		}
	}

	resp, err := d.client.CreateChatCompletion(ctx, openai.ChatCompletionRequest{
		Model:    d.model,
		Messages: openaiMessages,
	})
	if err != nil {
		// Try fallback to OpenAI if DeepSeek fails
		fmt.Printf("⚠️  DeepSeek failed: %v. Trying OpenAI fallback...\n", err)
		openaiClient, err := NewOpenAILLM()
		if err != nil {
			return "", fmt.Errorf("deepseek failed (%v) and OpenAI fallback failed (%v)", err, err)
		}
		return openaiClient.Generate(ctx, messages)
	}

	if len(resp.Choices) == 0 {
		// Try fallback to OpenAI if DeepSeek returns no choices
		fmt.Printf("⚠️  DeepSeek returned no choices. Trying OpenAI fallback...\n")
		openaiClient, err := NewOpenAILLM()
		if err != nil {
			return "", fmt.Errorf("deepseek returned no choices and OpenAI fallback failed (%v)", err)
		}
		return openaiClient.Generate(ctx, messages)
	}

	return resp.Choices[0].Message.Content, nil
}

func (d *DeepSeekLLM) Answer(ctx stdcontext.Context, question string, retrievedChunks []ingestion.Chunk, context context.ContextType) (string, error) {
	var prompt string

	if len(retrievedChunks) > 0 {
		// Build context from retrieved chunks
		var contextText string
		for _, chunk := range retrievedChunks {
			contextText += chunk.Text + "\n\n"
		}

		prompt = "Based on the following information, please answer the question:\n\n" +
			"Context:\n" + contextText + "\n\n" +
			"Question: " + question + "\n\n" +
			"Answer:"
	} else {
		prompt = "Please answer the following question:\n\n" +
			"Question: " + question + "\n\n" +
			"Answer:"
	}

	resp, err := d.client.CreateChatCompletion(ctx, openai.ChatCompletionRequest{
		Model: d.model,
		Messages: []openai.ChatCompletionMessage{
			{Role: "system", Content: groundedAnswerSystem},
			{Role: "user", Content: prompt},
		},
	})
	if err != nil {
		// Try fallback to OpenAI if DeepSeek fails
		fmt.Printf("⚠️  DeepSeek failed during answer: %v. Trying OpenAI fallback...\n", err)
		openaiClient, err := NewOpenAILLM()
		if err != nil {
			return "", fmt.Errorf("deepseek failed (%v) and OpenAI fallback failed (%v)", err, err)
		}
		return openaiClient.Answer(ctx, question, retrievedChunks, context)
	}

	if len(resp.Choices) == 0 {
		// Try fallback to OpenAI if DeepSeek returns no choices
		fmt.Printf("⚠️  DeepSeek returned no choices during answer. Trying OpenAI fallback...\n")
		openaiClient, err := NewOpenAILLM()
		if err != nil {
			return "", fmt.Errorf("deepseek returned no choices and OpenAI fallback failed (%v)", err)
		}
		return openaiClient.Answer(ctx, question, retrievedChunks, context)
	}

	return resp.Choices[0].Message.Content, nil
}

func (d *DeepSeekLLM) Edit(ctx stdcontext.Context, inputText string, instruction string, retrievedChunks []ingestion.Chunk, context context.ContextType) (string, error) {
	var prompt string

	if len(retrievedChunks) > 0 {
		// Build context from retrieved chunks
		var contextText string
		for _, chunk := range retrievedChunks {
			contextText += chunk.Text + "\n\n"
		}

		prompt = "Using the provided context and following the instruction, please edit the input text:\n\n" +
			"Context:\n" + contextText + "\n\n" +
			"Instruction: " + instruction + "\n\n" +
			"Input Text: " + inputText + "\n\n" +
			"Edited Text:"
	} else {
		prompt = "Please edit the following text according to the instruction:\n\n" +
			"Instruction: " + instruction + "\n\n" +
			"Input Text: " + inputText + "\n\n" +
			"Edited Text:"
	}

	resp, err := d.client.CreateChatCompletion(ctx, openai.ChatCompletionRequest{
		Model: d.model,
		Messages: []openai.ChatCompletionMessage{
			{Role: "system", Content: groundedEditSystem},
			{Role: "user", Content: prompt},
		},
	})
	if err != nil {
		// Try fallback to OpenAI if DeepSeek fails
		fmt.Printf("⚠️  DeepSeek failed during edit: %v. Trying OpenAI fallback...\n", err)
		openaiClient, err := NewOpenAILLM()
		if err != nil {
			return "", fmt.Errorf("deepseek failed (%v) and OpenAI fallback failed (%v)", err, err)
		}
		return openaiClient.Edit(ctx, inputText, instruction, retrievedChunks, context)
	}

	if len(resp.Choices) == 0 {
		// Try fallback to OpenAI if DeepSeek returns no choices
		fmt.Printf("⚠️  DeepSeek returned no choices during edit. Trying OpenAI fallback...\n")
		openaiClient, err := NewOpenAILLM()
		if err != nil {
			return "", fmt.Errorf("deepseek returned no choices and OpenAI fallback failed (%v)", err)
		}
		return openaiClient.Edit(ctx, inputText, instruction, retrievedChunks, context)
	}

	return resp.Choices[0].Message.Content, nil
}
